blocks/bricks/wrappers.py

```diff
@@ -2,40 +2,91 @@
 from theano import tensor
 
 from blocks.bricks.base import Brick, application
+from blocks.utils import dict_subset, pack
 
 
-class As2D(Brick):
-    """Wraps a 2D application around reshapes to accomodate N-D tensors.
+class WithExtraDims(Brick):
```

Comment:
Hi, 

I guess we are hitting two extremes - the code I pasted has a mask argument which is not needed if you allow Softmax to return arrays. If you get rid of the masks and just let it operate on a chosen mask it will become simpler. In fact in most places you replace axis=1 by axis=self.axis. The only non-trivial change is the dimshuffle for categorical cross-entropy.

I needed the axis awareness because I implemented an FF conv net for speech.

Also, the current implementation of SoftMax as an activation is simply misleading - the documentation says that it is element-wise, which is not true and you actually have to read the code to know on which axis it operates. So the implementation is long, because I had to copy the documentation.

Then, blocks should supply the users with good implementations of common use-cases. If even we need to handle n-dimensional arrays in softmax why not simply add it?